tests/runtime/model_management/test_model_manager.py

```diff
@@ -34,370 +38,739 @@
+                manager_one.loaded_models.get(model_id)
+                and manager_one.loaded_models[model_id].model() is model_one
             )
-        self.assertEqual(context.exception.status_code, grpc.StatusCode.NOT_FOUND)
-
-    # ****************************** Unit Tests ****************************** #
-    # These tests patch in mocks for the manager's dependencies, to test its code in isolation
 
-    def test_load_model(self):
-        """Test to make sure that given valid input, the model manager gives a happy response
-        when we tried to load in a model (model size > 0 or 0 if the model size will be computed
-        at a later time)."""
-        mock_loader = MagicMock()
-        mock_sizer = MagicMock()
+            # Make sure the same model is available on the second manager
+            assert manager_two.loaded_models.get(model_id) is None
+            assert manager_two.retrieve_model(model_id)
+            assert manager_two.loaded_models.get(model_id)
+
+            # Remove the model from the local dir and trigger both managers to
+            # sync their local model dirs to make sure it gets removed
+            shutil.rmtree(model_one_path)
+            manager_one.sync_local_models(wait=True)
+            manager_two.sync_local_models(wait=True)
+            with pytest.raises(CaikitRuntimeException) as context:
+                manager_one.retrieve_model(model_id)
+            with pytest.raises(CaikitRuntimeException) as context:
+                manager_two.retrieve_model(model_id)
+
+
+def test_model_manager_disk_caching_periodic_sync(good_model_path):
+    """Make sure that when using disk caching, the manager periodically syncs
+    its loaded models based on their presence in the cache
+    """
+    purge_period = 0.001
+    with TemporaryDirectory() as cache_dir:
+        with non_singleton_model_managers(
+            2,
+            {
+                "runtime": {
+                    "local_models_dir": cache_dir,
+                    "lazy_load_local_models": True,
+                    "lazy_load_poll_period_seconds": purge_period,
+                },
+            },
+            "merge",
+        ) as managers:
+            manager_one, manager_two = managers
+
+            # Load the model on the first manager and get it loaded into the
+            # the first by fetching it
+            model_id = random_test_id()
+            model_cache_path = os.path.join(cache_dir, model_id)
+            assert not os.path.exists(model_cache_path)
+            shutil.copytree(good_model_path, model_cache_path)
+            assert manager_one.retrieve_model(model_id)
+            assert model_id in manager_one.loaded_models
+
+            # Wait for the sync period and make sure it's available in the
+            # second manager
+            start = time.time()
+            mgr_two_loaded = False
+            while (time.time() - start) < (purge_period * 1000):
+                time.sleep(purge_period)
```

Comment:
sleeeeeps in tests :((((

I guess this one isn't too egregious, though if it ever starts failing it'll kill a second waiting to do so. I don't think there's really a better way to do this though unless you were to explicitly invoke the second model-manager's `sync`, which I guess defeats the purpose of checking that it is periodically syncing

Reply:
Heh, I was thinking of _exactly_ this comment when I put this loop in rather than just a static sleep. I figured a tight loop that knows when the condition is met was better!